File: iOS/Kalorie/Features/Moderation/ModerationReportsViewModel.swift
//
//  ModerationReportsViewModel.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

final class ModerationReportsViewModel: ObservableObject {

    // MARK: - Types

    struct ReportGroup: Identifiable {
        let barcode: String
        let itemName: String?
        let reports: [FoodItemReportDomain]

        var id: String { barcode }
    }

    // MARK: - Properties

    @Published private(set) var state: LoadingState<Void> = .idle
    @Published private(set) var groups: [ReportGroup] = []
    @Published var alertItem: AlertItem?

    private let fetchFoodItemReports: any FetchFoodItemReportsUseCaseProtocol
    private let fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    private let deleteFoodItemReport: any DeleteFoodItemReportUseCaseProtocol

    // MARK: - Init

    init(
        fetchFoodItemReports: any FetchFoodItemReportsUseCaseProtocol,
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol,
        deleteFoodItemReport: any DeleteFoodItemReportUseCaseProtocol
    ) {
        self.fetchFoodItemReports = fetchFoodItemReports
        self.fetchFoodItemByBarcode = fetchFoodItemByBarcode
        self.deleteFoodItemReport = deleteFoodItemReport
    }

    // MARK: - Functions

    @MainActor
    func onAppear() async {
        state = .loading
        defer { state = .loaded }
        do {
            let reports = try await fetchFoodItemReports()
            groups = await Self.grouped(reports, fetchFoodItemByBarcode: fetchFoodItemByBarcode)
        } catch {
            Log.error(error, category: Constants.LogCategory.moderation)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }

    @MainActor
    func onRefresh() async {
        await onAppear()
    }

    @MainActor
    func onResolveTapped(_ group: ReportGroup) async {
        var remainingReports: [FoodItemReportDomain] = []
        for report in group.reports {
            do {
                try await deleteFoodItemReport(barcode: report.barcode, reportedBy: report.reportedBy)
            } catch {
                Log.error(error, category: Constants.LogCategory.moderation)
                remainingReports.append(report)
            }
        }
        guard !remainingReports.isEmpty else {
            groups.removeAll { $0.barcode == group.barcode }
            return
        }
        if let index = groups.firstIndex(where: { $0.barcode == group.barcode }) {
            groups[index] = ReportGroup(barcode: group.barcode, itemName: group.itemName, reports: remainingReports)
        }
        alertItem = AlertItem(title: L10n.Common.errorUnknown)
    }

    // MARK: - Private

    private static func grouped(
        _ reports: [FoodItemReportDomain],
        fetchFoodItemByBarcode: any FetchFoodItemByBarcodeUseCaseProtocol
    ) async -> [ReportGroup] {
        let barcodes = Array(Set(reports.map(\.barcode)))
        let items = (try? await fetchFoodItemByBarcode(barcodes: barcodes)) ?? []
        let namesByBarcode = Dictionary(items.map { ($0.id, $0.displayName) }) { first, _ in first }
        let reportsByBarcode = Dictionary(grouping: reports, by: \.barcode)
        return reportsByBarcode
            .map { barcode, reports in
                ReportGroup(
                    barcode: barcode,
                    itemName: namesByBarcode[barcode],
                    reports: reports.sorted { $0.reportedAt > $1.reportedAt }
                )
            }
            .sorted { lhs, rhs in
                if lhs.reports.count != rhs.reports.count { return lhs.reports.count > rhs.reports.count }
                let lhsLatest = lhs.reports.first?.reportedAt ?? .distantPast
                let rhsLatest = rhs.reports.first?.reportedAt ?? .distantPast
                return lhsLatest > rhsLatest
            }
    }
}

File: Android/app/src/main/java/antoni/kalorie/features/moderation/ModerationReportsViewModel.kt
package antoni.kalorie.features.moderation

import androidx.lifecycle.ViewModel
import antoni.kalorie.R
import antoni.kalorie.core.models.FoodItemReportDomain
import antoni.kalorie.core.models.displayName
import antoni.kalorie.core.usecases.DeleteFoodItemReportUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodItemByBarcodeUseCaseProtocol
import antoni.kalorie.core.usecases.FetchFoodItemReportsUseCaseProtocol
import antoni.kalorie.core.utils.AlertItem
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.core.utils.LoadingState
import antoni.kalorie.core.utils.Log
import java.time.Instant
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

data class ReportGroup(
    val barcode: String,
    val itemName: String?,
    val reports: List<FoodItemReportDomain>,
)

class ModerationReportsViewModel(
    private val fetchFoodItemReports: FetchFoodItemReportsUseCaseProtocol,
    private val fetchFoodItemByBarcode: FetchFoodItemByBarcodeUseCaseProtocol,
    private val deleteFoodItemReport: DeleteFoodItemReportUseCaseProtocol,
) : ViewModel() {

    // MARK: - Properties

    private val _state = MutableStateFlow<LoadingState<Unit>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<Unit>> = _state
    private val _groups = MutableStateFlow<List<ReportGroup>>(emptyList())
    val groups: StateFlow<List<ReportGroup>> = _groups
    val alertItem = MutableStateFlow<AlertItem?>(null)

    // MARK: - Functions

    suspend fun onAppear() {
        _state.value = LoadingState.Loading
        try {
            val reports = fetchFoodItemReports()
            _groups.value = grouped(reports)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.MODERATION)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            _state.value = LoadingState.loaded
        }
    }

    suspend fun onRefresh() {
        onAppear()
    }

    suspend fun onResolveTapped(group: ReportGroup) {
        val remainingReports = mutableListOf<FoodItemReportDomain>()
        for (report in group.reports) {
            try {
                deleteFoodItemReport(report.barcode, report.reportedBy)
            } catch (error: CancellationException) {
                throw error
            } catch (error: Exception) {
                Log.error(error, Constants.LogCategory.MODERATION)
                remainingReports += report
            }
        }
        if (remainingReports.isEmpty()) {
            _groups.value = _groups.value.filterNot { it.barcode == group.barcode }
            return
        }
        _groups.value = _groups.value.map { existing ->
            if (existing.barcode == group.barcode) existing.copy(reports = remainingReports) else existing
        }
        alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
    }

    // MARK: - Private

    private suspend fun grouped(reports: List<FoodItemReportDomain>): List<ReportGroup> {
        val barcodes = reports.map { it.barcode }.distinct()
        val items = try {
            fetchFoodItemByBarcode(barcodes)
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.MODERATION)
            emptyList()
        }
        val namesByBarcode = buildMap { items.forEach { putIfAbsent(it.id, it.displayName) } }
        return reports.groupBy { it.barcode }
            .map { (barcode, barcodeReports) ->
                ReportGroup(
                    barcode = barcode,
                    itemName = namesByBarcode[barcode],
                    reports = barcodeReports.sortedByDescending { it.reportedAt },
                )
            }
            .sortedWith(
                compareByDescending<ReportGroup> { it.reports.size }
                    .thenByDescending { it.reports.firstOrNull()?.reportedAt ?: Instant.MIN },
            )
    }
}
